guest-tools/td_virsh_tool.sh

```diff
@@ -0,0 +1,224 @@
+#
+# Note this script supports running multiple VMs simultaneously,
+# either by passing the -n flag or by running the script multiple times.
+#
+
+##
+# Global constants
+#
+DOMAIN_PREFIX="td_guest"
+SCRIPT_DIR=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)
+BASE_IMG=${TD_IMG:-${SCRIPT_DIR}/image/tdx-guest-ubuntu-23.10.qcow2}
+XML_TEMPLATE=${XML_TEMPLATE:-${SCRIPT_DIR}/td_guest.xml.template}
+MAX_DOMAINS=${MAX_DOMAINS:-"20"}
+
+##
+# Global variables
+#
+n_instances=1
+domain=${DOMAIN_PREFIX}-"1"
+overlay_image_path=/tmp/overlay.${domain}.qcow2
+base_img_path=""
+xml_template_path=""
+domain_to_clean=""
+
+main() {
+    parse_params "$@"
+
+    if [ "${domain_to_clean}" = "all" ]; then
+        clean_all
+    elif [ ! -z "${domain_to_clean}" ]; then
+        clean_one
+    fi
+
+    set_input_paths
+
+    set -e
+    trap "on_exit" EXIT
+
+    for ((i = 0; i < ${n_instances}; i++)); do
+        set_domain
+        create_overlay_image
+        create_domain_xml
+        boot_vm
+        echo_ssh_cmd
+    done
+}
+
+usage() {
+    cat <<EOM
+Usage: ./$(basename "${BASH_SOURCE[0]}") [-h] [-c D] [-n N]
+
+Run TD VMs managed by libvirt.
+
+Available options:
+
+-h,   --help            Print this help and exit
+-c D, --clean D         Clean domain with name D (use "all" to clean all)
+-n N, --n_instances N   Launch N instances
+
+Environment variables:
+
+TD_IMG         TDX image path (default: ./image/tdx-guest-ubuntu-23.10.qcow2)
+XML_TEMPLATE   Path to virsh guest XML template (default: ./td_guest.xml.template)
+MAX_DOMAINS    Maximum allowed domains running (default: 20)
+EOM
+}
+
+parse_params() {
+    while :; do
+        case "${1-}" in
+        -h | --help)
+            usage
+            exit 0
+            ;;
+        -n | --n_instances)
+            n_instances="${2-}"
+            if [ -z ${n_instances} ]; then
+                echo "Please pass instance count after -n"
+                exit 1
+            fi
+            shift
+            ;;
+        -c | --clean)
+            domain_to_clean="${2-}"
+            if [ -z ${domain_to_clean} ]; then
+                echo "Please pass name of domain to destroy (or \"all\" to destroy all)"
+                exit 1
+            fi
+            shift
+            ;;
+        -?*)
+            echo "Unknown option: $1"
+            usage
+            exit 1
+            ;;
+        *) break ;;
+        esac
+        shift
+    done
+}
+
+set_input_paths() {
+    base_img_path=$(realpath ${BASE_IMG})
+    if [ $? -ne 0 ] || [ ! -f ${base_img_path} ]; then
+        echo "Set base TD qcow2 image path with TD_IMG"
+        exit 1
+    fi
+    xml_template_path=$(realpath ${XML_TEMPLATE})
+    if [ $? -ne 0 ] || [ ! -f ${xml_template_path} ]; then
+        echo "Set libvirt guest XML template path with XML_TEMPLATE"
+        exit 1
+    fi
+}
+
+set_domain() {
+    local domain_id=1
+    while $(virsh list --state-running --name |
+        grep -q "\-${domain_id}"); do
```

Comment:
we are using a incremental counter for domain identification, i think this is hard to manage, we have to find the free id each time, using something like UUID is more robust i would think, it is kind of random number (with a very low probability of collision) when you want to generate an ID for a new domain, just generate an UUID. This will also be more robust wrt to domains that user created that are not related to TDX

Reply:
One challenge of using UUID is however to handle nicely the clean up of qcow image

Reply:
If we want to stick with the counter, perhaps add a DOMAIN_PREFIX in the "grep" to make it more  robust

Reply:
I like the UUID suggestion. It also occurred to me that this will be more robust in case multiple users are using the script on the same system at once. 

One issue I see is that it looks like domains already have UUIDs assigned:

```bash
$ virsh domuuid td_guest-1
8247b776-b3c9-4be4-8994-be440b934373
```

To avoid confusion, maybe I could append a random string (making use of `/dev/urandom`) or a date string on the end of the domain name.

* random string: `td_guest-j8uw4sNkehoZMIs`
* date-based (with millisecond resolution): `td_guest-20240322:093122434`

Reply:
I meant that indeed, append UUID to the domain name

Reply:
Done. Note that I'm embedding a random string in the qcow2 overlay images for each guest to ensure uniqueness. Then I use `virsh dumpxml` to grab the qcow2 path for cleanup. 

```bash
$ ./td_virsh_tool.sh 
Domain td_guest-468ee3e2-cbc1-4c8e-ab69-51f308fa2211 running with vsock CID: 3, ssh -p 33593 root@localhost

$ ./td_virsh_tool.sh -c td_guest-468ee3e2-cbc1-4c8e-ab69-51f308fa2211
Destroying domain td_guest-468ee3e2-cbc1-4c8e-ab69-51f308fa2211.
Waiting for VM to shutdown ...
```